subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/GroovyRecompilationSpecProvider.java

```diff
@@ -0,0 +1,181 @@
+            for (String staleClass : classes) {
+                String path = staleClass.replaceAll("\\.", "/");
+                classesToDelete.include(path.concat(".class"));
+            }
+        }
+    }
+
+    private void includePreviousCompilationOutputOnClasspath(JavaCompileSpec spec) {
+        List<File> classpath = Lists.newArrayList(spec.getCompileClasspath());
+        File destinationDir = spec.getDestinationDir();
+        classpath.add(destinationDir);
+        spec.setCompileClasspath(classpath);
+    }
+
+    private String relativize(File sourceFile) {
+        List<File> dirs = sourceDirs.getSourceRoots();
+        for (File sourceDir : dirs) {
+            if (sourceFile.getAbsolutePath().startsWith(sourceDir.getAbsolutePath())) {
+                return RelativePathUtil.relativePath(sourceDir, sourceFile);
+            }
+        }
+        throw new IllegalStateException("Not found " + sourceFile + " in source dirs: " + dirs);
+    }
+
+//    @SuppressWarnings("fallthrough")
+//    private void processClasspathChanges(CurrentCompilation current, PreviousCompilation previous, RecompilationSpec spec) {
+//        ClasspathEntryChangeProcessor classpathEntryChangeProcessor = new ClasspathEntryChangeProcessor(current.getClasspathSnapshot(), previous);
+//        Set<File> transformedClasspathEntryDetectionSet = Sets.newHashSet();
+//        for (FileChange fileChange : classpathChanges) {
+//            switch (fileChange.getChangeType()) {
+//                case ADDED:
+//                case REMOVED:
+//                    if (!transformedClasspathEntryDetectionSet.add(fileChange.getFile())) {
+//                        spec.setFullRebuildCause("Classpath has been changed", null);
+//                        return;
+//                    }
+//                case MODIFIED:
+//                    classpathEntryChangeProcessor.processChange((DefaultFileChange) fileChange, spec);
+//                    break;
+//            }
+//        }
+//    }
+
+    private void processOtherChanges(CurrentCompilation current, PreviousCompilation previous, RecompilationSpec spec) {
+        if (spec.getFullRebuildCause() != null) {
+            return;
+        }
+        SourceFileChangeProcessor sourceFileChangeProcessor = new SourceFileChangeProcessor(previous, sourceFileClassNameConverter);
+        for (FileChange fileChange : sourceChanges) {
+            if (spec.getFullRebuildCause() != null) {
+                return;
+            }
+            spec.getFilesToCompile().add(fileChange.getFile());
+            sourceFileChangeProcessor.processChange((DefaultFileChange) fileChange, spec);
+        }
+
+        for (String className : spec.getClassesToCompile()) {
+            if (spec.getFullRebuildCause() != null) {
+                return;
+            }
+
+            Optional<File> sourceFile = sourceFileClassNameConverter.getFile(className);
```

Comment:
I'm a little concerned about the performance when inverse looking for a file by class name.

Reply:
And you should :)

The performance problem is easily solved by constructing the reverse map from fqcn to source file.